Tiltfile

```diff
@@ -0,0 +1,181 @@
+# Language: Go
+###
+docker_build_with_restart('productcatalogservice',
+    context='./src/productcatalogservice',
+    entrypoint='/src/server',
+    live_update=[
+        sync('./src/productcatalogservice', '/src'),
+    ],
+)
+
+###
+# Service: currencyservice
+# Desc: Converts one money amount to another currency.
+#       Uses real values fetched from European Central Bank.
+#       It's the highest QPS service
+# Language: Node.js
+###
+docker_build_with_restart('currencyservice',
+    context='./src/currencyservice',
+    entrypoint='node /usr/src/app/server.js',
+    live_update=[
+        sync('./src/currencyservice', '/usr/src/app'),
+        run('cd /usr/src/app && npm install', trigger=['./src/currencyservice/package.json']),
+    ]
+)
+
+###
+# Service: paymentservice
+# Desc: Charges the given credit card info (mock) with the given amount
+#       and returns a transaction ID
+# Language: Node.js
+###
+docker_build_with_restart('paymentservice',
+    context='./src/paymentservice',
+    entrypoint='node /usr/src/app/index.js',
+    live_update=[
+        sync('./src/paymentservice', '/usr/src/app'),
+        run('cd /usr/src/app && npm install', trigger=['./src/paymentservice/package.json']),
+    ]
+)
+
+###
+# Service: shippingservice
+# Desc: Gives shipping cost estimates based on the shopping cart.
+#       Ships items to the given address (mock)
+# Language: Go
+###
+docker_build_with_restart('shippingservice',
+    context='./src/shippingservice',
+    entrypoint='/src/shippingservice',
+    live_update=[
+        sync('./src/shippingservice', '/src'),
+    ],
+)
+
+###
+# Service: emailservice
+# Desc: Sends users an order confirmation email (mock)
+# Language: Python
+###
+docker_build('emailservice',
+    context='./src/emailservice',
+    live_update=[
+        sync('./src/emailservice', '/email_server'),
+        run('cd /email_server && pip install -r requirements.txt',
+            trigger='./src/emailservice/requirements.txt'),
+    ],
+)
+
+###
+# Service: checkoutservice
+# Desc: Retrieves user cart, prepares order and orchestrates the payment,
+#       shipping and the email notification
+# Language: Go
+###
+docker_build_with_restart('checkoutservice',
+    context='./src/checkoutservice',
+    entrypoint='/src/checkoutservice',
+    live_update=[
+        sync('./src/checkoutservice', '/src'),
+    ],
+)
+
+###
+# Service: recommendationservice
+# Desc: Recommends other products based on what's given in the cart
+# Language: Python
+###
+docker_build('recommendationservice',
+    context='./src/recommendationservice',
+    live_update=[
+        sync('./src/recommendationservice', '/recommendationservice'),
+        run('cd /recommendationservice && pip install -r requirements.txt',
+            trigger='./src/recommendationservice/requirements.txt'),
+    ],
+)
+
+###
+# Service: adservice
+# Desc: Provides text ads based on given context words
+# Language: Java
+###
+docker_build_with_restart('adservice',
```

Comment:
**Question:**
Sorry, I've never used Tilt before, so I have some beginner questions.
I see that each service is using `docker_build_with_restart` or `docker_build`.
1. Does that mean each service is simply re-`docker build`-ing whenever a code changes occurs in that service (during development)?

2. If so, is this Tilt configuration different from the [existing `skaffold` configuration](https://github.com/GoogleCloudPlatform/microservices-demo/blob/main/skaffold.yaml)? The existing [`skaffold dev` option](https://github.com/GoogleCloudPlatform/microservices-demo/blob/main/docs/development-guide.md#option-2---local-cluster) rebuilds each service is a code changes occurs in that service's source code or Dockerfile.

I am hesitant to provide Online Boutique users with 2 options for local development because of the maintenance cost.
I think Online Boutique should be opinionated and support a single, consistent user journey for local development (i.e., `skaffold dev`).

Reply:
The difference between Skaffold and Tilt are on the official website https://docs.tilt.dev/skaffold.html.

TL;TR
Tilt's user interface allows you to easily see the status of your services and quickly identify any errors. You can view logs for individual services or see a global log of all activity. Tilt's configuration is written in Starlark, a simplified version of Python, which makes it easy to write simple configurations and allows for more complex configurations as well and Tilt have multiple extensions https://github.com/tilt-dev/tilt-extensions.

And you're right, the `docker_build_with_restart` and `docker_build` methods are used to build and rebuild Docker images for services in a Tilt project. These methods are similar to the Skaffold `dev` option in that they both rebuild services when code changes occur in the service's source code or Dockerfile.

The Tilt `docker_build_with_restart` method is used for services that need to restart their entrypoint/command (https://github.com/tilt-dev/tilt-extensions/tree/master/restart_process) when changes occur to the source code of that service or Docker file.

--

However, it's important to keep in mind that different users may have different preferences and needs when it comes to local development. As a result, it may be worth considering providing users with a choice of tools, along with clear documentation and guidance on how to use each tool. This can give users the flexibility to choose the right tool for their needs.